### DISCUSSION TITLE
How to chain two methods togther that use Aff

### DISCUSSION THREAD

**Discussion by dirkil** [louthy/language-ext#1281] [General]
I have difficulty in getting method calls chained together that use `Aff` and can fail. Let's say I have an interface with a method that calls a Rest API and returns `Aff<Unit>`.  Now I want to write another method that uses that interface method and embeds it in some error handling / logging. So basically, I have the following code (does not compile):

```
public interface IAccessClient
{
    Aff<Unit> DeleteClient(Guid clientId);
}

public record InputData(string Name)
{
    public bool IsValid() => true;
}

public class ChainTest
{
    private IAccessClient _accessClient;

    private Aff<InputData> Delete(Guid clientId, InputData inputData)
    {
        return AffMaybe<InputData>(async () =>
        {
            if (inputData.IsValid())
            {
                LogSomething();
                return _accessClient.DeleteClient(clientId).Match(
                    Succ: _ =>
                    {
                        LogSomething();
                        return Fin<InputData>.Succ(inputData);
                    },
                    Fail: error => error);
            }

            return Fin<InputData>.Succ(inputData);
        });
    }

    private void LogSomething()
    {
        // Execute synchronous logging steps
    }
```

In essence, I want to call `ChainTest.Delete().Run()`. It's probably a simple scenario but I don't get it to compile. Could someone please enlighten me how to chain these two calls? Any help is very much appreciated.

Regards,
Dirk.

**Comment by louthy** [comment]
```c#
Aff<X> effect = from result1 in YourFirstAffFunction()    // Aff effect
                from result2 in YourSecondAffFunction()   // Aff effect
                select UseBothValues(result1, result2);   // Regular, non-Aff returning function

Fin<X> result = await effect.Run();
````
This is called monadic binding.  All monadic types support this method of chaining.  In fact, you can think of monads as programmable semi-colons (the side-effect of the monadic flavour is run between the lines of your code).

If you weren't using LINQ, then the code above would be written like so:

```c#
Aff<X> effect = YourFirstAffFunction().Bind(result1 =>
                    YourSecondAffFunction().Map(result2 =>
                        UseBothValues(result1, result2)));

Fin<X> result = await effect.Run();
````
Which obviously isn't as elegant, so it's worth using LINQ when programming with monadic types.

This can be a little bit of a headfuk when you first start this style of coding, so there's a couple of wiki pages that should help:

* [What is LINQ really?](https://github.com/louthy/language-ext/wiki/Thinking-Functionally:-What-is-LINQ-really%3F)
* [Does C# Dream of Electric Monads?](https://github.com/louthy/language-ext/wiki/Does-C%23-Dream-Of-Electric-Monads%3F)

One of the key aspects is try to never call `Run`.  Or, ideally only call it at the 'edges' of your program.  The bit were you need to realise the effect.  Examples of 'edges' are:

* The program's `Main` function
* A web-request (even then, ideally your web-framework would know what an `Aff` is and so you never need to call `Run`.

If you look at thew [`EffectsExamples`](https://github.com/louthy/language-ext/tree/main/Samples/EffectsExamples) sample app, you'll notice that `Aff` is used everywhere, but `Run` only gets called in one place: [`Main`](https://github.com/louthy/language-ext/blob/main/Samples/EffectsExamples/Program.cs)

Here's your `ChainExample` made to leverage the power of the monads:

```c#
public class ChainTest
{
    IAccessClient _accessClient;
    
    public Aff<InputData> Delete(Guid clientId, InputData inputData) =>
        from _1 in guard(inputData.IsValid(), Error.New("Input data isn't valid")) 
        from _2 in LogSomething("upfront logging")
        from _3 in _accessClient.DeleteClient(clientId)
        from _4 in LogSomething("additiona logging")
        select inputData;  

    Aff<Unit> LogSomething(string msg) =>
        Aff<Unit>(() => {
            // Do your logging here
            return default;
        });
}

```
Hopefully it should be clearer now why they're so powerful, they get rid of soooo much boilerplate.

> btw, the `_1`, `_2`, `_3`, `_4` are just a convention I used for expressions where I discard the result.  You could just as easily use any var names you like.  Hopefully, Microsoft will implement the discard pattern for LINQ one day!

**Reply by dirkil** [reply]
Thank you very much for your detailed answer; that was very helpful. I have never looked deeply into the LINQ-style since I preferred the fluent syntax but having access to all the return values (_1, _2, etc.) in the lines further down the pipeline is a huge advantage. I like it!

One more thing to ask in that context: how can I setup the pipeline when the pipeline should only be executed when `IsValid() == true`? I don't want it to be regarded as an Error if `IsValid() == false`, it just shouldn't execute the pipeline and should just return `inputData` instead.

Maybe it's easier to imagine, if there is a function `NeedsDeletion()`. Since I want to avoid if-statements if possible, the best I could come up with is that:

```csharp
private Aff<InputData> Delete(Guid clientId, InputData inputData) =>
        inputData.NeedsDeletion()
            ?   from _1 in LogInformation("some text")
                from _2 in _accessClient.DeleteClient(clientId)
                from _3 in LogInformation("some other text")
                select inputData
            : inputData; // -> this does not compile! How can specify input as a return value?

```
How can I handle the case that it just returns `inputData` when `inputData.NeedsDeletion() == false`? Or maybe is there even a better way of doing that?

**Reply by louthy** [reply]
> inputData; // -> this does not compile! How can specify input as a return value?

Remember, everything in the monadic world lives in the 'lifted space'.  'Lifted' means that regular values and functions get wrapped up in the monad.

So, a value from `A`, when _lifted_ is `M<A>` (where the `M` is the type of monad).  To lift any value into the monadic space you need to know what its `A → M<A>` function is (also known as its _pure_ function, or _return_ function, or even just its constructor).  For `Aff` it's `SuccessAff(...)`, for `Option` it's `Some(...)`, for `Either` it's `Right(...)`, etc.  In future versions of language-ext you'll be able to use `Pure(value)` for all monadic types, which makes things a little easier.

So, change your code to this:
```c#
private Aff<InputData> Delete(Guid clientId, InputData inputData) =>
        inputData.NeedsDeletion()
            ?   from _1 in LogInformation("some text")
                from _2 in _accessClient.DeleteClient(clientId)
                from _3 in LogInformation("some other text")
                select inputData
            : SuccessAff(inputData);
```
And everything will magically work.

You don't get to mix the lower pure space and the lifted `Aff` space.  To work with regular-values, lift them first.  To work with regular functions use `Map(f)` where `f` is the function.  It can b thought of as lifting the function into the monad.

Why is it called 'lifting'?  Because diagrams of this usually have the monad space _above_ the _pure_ space:
```
              Map(A → B)
     M<A> ------------------- M<B>
      |                        |
     Pure                     Pure
      |                        |
      A ---------------------- B
                A → B
```
There are other ways of doing this also:

```c#
    public Aff<InputData> SafeDelete(Guid clientId, InputData inputData) =>
        Delete(clientId, inputData) | SuccessAff(inputData);

    public Aff<InputData> Delete(Guid clientId, InputData inputData) =>
        from _1 in guard(inputData.IsValid(), Error.New("Input data isn't valid")) 
        from _2 in LogSomething("upfront logging")
        from _3 in _accessClient.DeleteClient(clientId)
        from _4 in LogSomething("additiona logging")
        select inputData; 
```
The `|` operator works a little bit like the `null` coalescing operator, in that if anything to the left of the operator fails, then the right is run, but the right is never run if the left-hand-side succeeds.

It's worth looking at the error handling section of the ['How To Deal With Side Effects' wiki article](https://github.com/louthy/language-ext/wiki/How-to-deal-with-side-effects#how-to-catch-errors-from-aff-or-eff-sub-expressions)

**Reply by dirkil** [reply]
This is really great stuff! I am really excited about this. I knew that I needed a Return function to lift it up and each monad must have one (otherwise it wouldn't be a monad) but I tried stuff like `Aff<InputData>.Succ(inputData)`. 

I am currently re-writing an existing loader and I am amazed how much shorter the code becomes with your LanguageExt and how much more readable it gets. 

If you don't mind, I have another question which is similar. I have a method that needs to process a list of items. Each call is async and could potentially fail. Even if the processing of an item fails it should carry on to process the other items. The result of all should be a list of processing results. The code looks something like this:

```csharp
public Aff<IEnumerable<ProcessResult>> Process(List<string> itemsToProcess)
    {
        return itemsToProcess.Map(ProcessItem);  // Does not compile!
    }
    
    private Aff<ProcessResult> ProcessItem(string itemId)
    {
        // Do some processing
        return LanguageExt.Aff<ProcessResult>.Success(new ProcessResult(itemId, true, None)); 
    }
    
    Aff<Unit> LogSomething(string msg) =>
        Aff<Unit>(() => {
            // Do your logging here
            return default;
        });

```

My problem is that `Process()` needs to return an `Aff<IEnumerable<ProcessResult>>` but it returns an `IEnumerable<Aff<ProcessResult>>`. So basically it needs to be switched. I thought I could use `Sequence()` for that but I couldn't get it to work. Is there an easy fix for that or how would you manage a set of results from a method that returns an `IEnumerable<Aff<T>>`?

**Reply by louthy** [reply]
I'm away for a week now (off to the Abu Dhabi Grand Prix, yay!).  So, it'll either have to wait, or someone else will help, or you'll work it out! 

Good luck!

**Reply by StefanBertels** [reply]
I think `.Sequence()` is what you want. Maybe try `Seq` or `Arr` instead of `List` / `IEnumerable` => `public Aff<Arr<ProcessResult>> Process(Arr<string> itemsToProcess)`...

**Reply by louthy** [reply]
SequenceParallel or SequenceSerial for asynchronous types

**Reply by dirkil** [reply]
Yes, SequenceParallel is the way to go. 

Thank you all for yourgreat support!

And Paul: have fun in Abu Dhabi! 🥇
